### PR TITLE
Add ClassicalSimulator for faster testing of non-quantum circuits

### PR DESCRIPTION
**Pull request by Strilanc**

### CODE REVIEW

**Comment by Strilanc** [comment]
Comes after https://github.com/ProjectQ-Framework/ProjectQ/pull/49

File: projectq/backends/_sim/_classical_simulator.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        Reads a group of bits as a little-endian integer.
+
+        Args:
+            qureg (projectq.types.Qureg):
+                The group of bits to read, in little-endian order.
+
+        Returns:
+            int: Little-endian register value.
+        """
+        return sum(self.read_bit(qureg[i]) << i
```
Comment (thomashaener) [inline projectq/backends/_sim/_classical_simulator.py]: I'd suggest using `result |= (value << i)` in a loop (it's almost 2x faster than summation)

File: projectq/backends/_sim/_classical_simulator.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+        if isinstance(cmd.gate, XGate):
+            assert len(cmd.qubits) == 1 and len(cmd.qubits[0]) == 1
+            target = cmd.qubits[0][0]
+            if all(self.read_bit(c) for c in cmd.control_qubits):
+                self.write_bit(target, not self.read_bit(target))
+            return
+
+        if isinstance(cmd.gate, BasicMathGate):
+            if all(self.read_bit(c) for c in cmd.control_qubits):
```
Comment (thomashaener) [inline projectq/backends/_sim/_classical_simulator.py]: For the controls I'd suggest building a mask and then compare `mask & state == mask`.

**Comment by thomashaener** [comment]
I added a few suggestions to improve performance. I think these will help when testing large circuits (for, e.g., Shor's algorithm). I will also add a C++ version as soon as I find time.

File: projectq/backends/_sim/_classical_simulator.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
+"""
```
Comment (damiansteiger) [inline projectq/backends/_sim/_classical_simulator.py]: Missing license header:

Every new file should start with the same Apache license boiler plate.

Comment (Strilanc) [inline projectq/backends/_sim/_classical_simulator.py]: Done.

File: projectq/backends/_sim/_classical_simulator.py (hunk 2 above)
Comment (Strilanc) [inline projectq/backends/_sim/_classical_simulator.py]: Done.

File: projectq/backends/_sim/_classical_simulator.py (hunk 1 above)
Comment (Strilanc) [inline projectq/backends/_sim/_classical_simulator.py]: Done.

**Comment by Strilanc** [comment]
Note: still comes after #49

**Comment by thomashaener** [comment]
Oh I think the issue was the PR title which is different from what the PR actually addresses: It fixes the case where there is an empty control register, rather than n=0. Controlling with n=0 works fine as long as there is no empty register.

File: projectq/backends/_sim/_classical_simulator.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            assert len(cmd.qubits) == 1 and len(cmd.qubits[0]) == 1
+            target = cmd.qubits[0][0]
+            if meets_controls:
+                self.write_bit(target, not self.read_bit(target))
+            return
+
+        if isinstance(cmd.gate, BasicMathGate):
+            if meets_controls:
+                ins = [self.read_register(reg) for reg in cmd.qubits]
+                outs = cmd.gate.get_math_function(None)(ins)
```
Comment (thomashaener) [inline projectq/backends/_sim/_classical_simulator.py]: get_math_function needs cmd.qubits (if users override its behavior such as in the AdvancedBasicMath gate you proposed).

Comment (Strilanc) [inline projectq/backends/_sim/_classical_simulator.py]: Done.

**Comment by thomashaener** [comment]
Great, thanks!

**Comment by damiansteiger** [comment]
👍

**Pull request merged**